### DISCUSSION TITLE
TCP/IP 인터넷 프로토콜 스위트

### DISCUSSION THREAD

**Discussion by seunghye218** [2024-woowacourse-study/level-interview#214] [Network]
TCP/IP 4계층

**Comment by le2sky** [comment]
인터넷 프로토콜 스위트는 인터넷에서 컴퓨터들이 서로 정보를 주고받는 데 쓰이는 통신 규약의 모음입니다.
이 중 TCP/IP가 가장 많이 사용되기 때문에 TCP/IP 프로토콜 스위트라고 불리기도 합니다.

TCP/IP 프로토콜 스위트는 4계층으로 나누어져있습니다.
데이터는 4계층에 걸쳐 만들어지고, 다시 이 4계층을 역으로 거쳐서 전달됩니다.
이 과정에서 각 단계마다 캡슐화와 역캡슐화가 이루어집니다. 
캡슐화는 데이터 전송 과정에서 상위 계층에서 하위 계층으로 이동할 때, 필요한 헤더가 추가되는 것을 의미합니다.
역캡슐화는 데이터 수신 과정에서 하위 계층에서 상위 계층으로 이동할 때, 추가된 헤더를 해석하고 행동을 취한 뒤에 제거하는 것을 의미합니다.

계층은 다음과 같이 이루어져있습니다.

1. 네트워크 액세스 계층: 
    - 데이터를 전기신호로 변환하고, 물리적 주소인 MAC 주소를 이용해 목적지 기기로 데이터를 전달하는 계층입니다. 
    - 대표적인 프로토콜로는 이더넷 프로토콜이 있습니다. 
    
2. 인터넷 계층 :
    - 패킷을 라우팅하는 계층입니다. 
    - 대표적인 프로토콜로 IP, ARP, ICMP 등이 있습니다. 
  
3. 전송 계층 :
    - 통신 단말간의 데이터 전송을 담당하는 계층입니다. 
    - 역캡슐화 과정에서 포트를 사용해 프로세스에게 데이터를 전달하기도 합니다. 
    - 대표적인 프로토콜로 TCP, UDP 등이 있습니다.

4. 애플리케이션 계층 :
    - 사용자와 소프트웨어 사이의 소통을 담당합니다.
    - 대표적인 프로토콜로는 HTTP, HTTPS, DNS, SSH, FTP, SMTP 등이 있습니다.

**Comment by J-I-H-O** [comment]
인터넷 프로토콜 스위트란 컴퓨터들이 정보를 주고 받는데 쓰이는 통신 규약의 모음입니다. 
TCP/IP 4계층은 각각의 역할과 책임에 따라 4개의 계층으로 나눈 것으로, 상업적이고 실무적으로 이용될 수 있도록 단순화한 모형입니다.

- **응용 계층 (Application Layer)**
    - 응용 프로그램들이 데이터를 교환하기 위해 사용되는 프로토콜입니다.
    - HTTP, FTP, SSH, Telnet, DNS, SMTP 등
    - 데이터 단위: Data, Message
- **전송 계층 (Transport Layer)**
    - 통신 노드 간 연결 제어 및 자료 송수신을 담당합니다.
        - 발신 및 수신 포트번호, 오류 검출 코드 등을 헤더에 추가합니다.
    - TCP, UDP, RTP, RTCP
    - 데이터 단위: Segment
- **인터넷 계층 (Internet Layer)**
    - 네트워크상 최종 목적지까지 정확하게 연결되도록 연결성을 제공합니다.
    - 단말을 구분하기 위해 논리적인 주소인 IP를 할당합니다.
        - 출발지와 목적지의 IP 등을 헤더에 추가합니다.
    - 라우팅 기능 처리하여 경로를 설정합니다.
    - IP, ARP, ICMP, RARP, OSPF
    - 데이터 단위: Packet
- **네트워크 연결 계층 (Network Access Layer)**
    - 물리적으로 데이터가 네트워크를 통해 어떻게 전송되는지 정의합니다.
    - 논리 주소(IP)가 아닌 물리 주소(MAC)를 참조해 장비 간 전송을 수행합니다. 이때 MAC 주소란, 컴퓨터의 하드웨어 주소를 말합니다.
    - 에러 검출과 패킷의 프레임화를 담당하며, 최종적으로 데이터 전송 전에 헤더에 MAC 주소와 오류 검출을 위한 정보를 헤더에 추가합니다.
    - 데이터 단위: Frame
